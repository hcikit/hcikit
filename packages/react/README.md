# @hcikit/react

This package is a package that allows you to create and run HCI experiments. Visit hcikit.github.io for more information.
