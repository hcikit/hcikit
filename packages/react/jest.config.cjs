/** @type {import('@ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  moduleNameMapper: {
    "@hcikit/workflow(.*)$": "<rootDir>/../workflow/src$1",
  },
  testMatch: ["./src/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[t]s?(x)"],
  modulePathIgnorePatterns: ["./dist/", "./node_modules/"],
  testPathIgnorePatterns: ["./dist/", "./node_modules/"],
  transformIgnorePatterns: ["./dist/", "/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/setupTests.ts"],
  collectCoverageFrom: [
    "src/**/*.{ts,tsx}",
    "!/node_modules/",
    "!src/setupTests.ts",
    "!src/**/*.stories.{js,ts,tsx,jsx}",
  ],
};
