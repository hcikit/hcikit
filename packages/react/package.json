{
  "name": "@hcikit/react",
  "version": "1.0.0-alpha.13",
  "description": "An experiment runner with a set of pre-built experiment components for running hci experiments",
  "author": "blainelewis1",
  "license": "MIT",
  "repository": "hcikit/hcikit",
  "module": "dist/index.js",
  "main": "dist/index.js",
  "type": "module",
  "keywords": [
    "react",
    "hci",
    "experiment"
  ],
  "types": "./dist/index.d.ts",
  "exports": {
    "import": "./dist/index.js",
    "default": "./dist/index.js"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "scripts": {
    "test": "NODE_OPTIONS=--experimental-vm-modules yarn jest --env=jsdom",
    "test:watch": "NODE_OPTIONS=--experimental-vm-modules yarn jest --watch --env=jsdom --colors --verbose",
    "start": "tsc --watch",
    "build": "tsc --build --verbose",
    "prepack": "yarn run build",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "peerDependencies": {
    "@mui/icons-material": "^5.5.0",
    "@mui/material": "^5.5.0",
    "@mui/styles": "^5.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@hcikit/workflow": "workspace:*",
    "core-js": "^3.26.1",
    "lodash-es": "^4.17.21",
    "prop-types": "^15.8.1",
    "react-error-boundary": "^3.1.4",
    "react-is": "^18.2.0",
    "react-markdown": "^8.0.3",
    "serialize-error": "^11.0.0",
    "styled-components": "^5.3.6"
  },
  "gitHead": "8d0f2af41121f70996e8b5803d006832b147fd86",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.25",
    "@types/styled-components": "^5.1.26",
    "@types/testing-library__jest-dom": "^5.14.5",
    "jest-diff": "^29.3.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "^5.0.1",
    "typescript": "^4.9.3"
  }
}
