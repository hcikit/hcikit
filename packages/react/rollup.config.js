import babel from "rollup-plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import resolve from "@rollup/plugin-node-resolve";
import pkg from "./package.json";
import postcss from "rollup-plugin-postcss";

export default {
  input: "src/index.js",
  output: [
    {
      file: pkg.main,
      format: "cjs",
      sourcemap: true,
    },
    {
      file: pkg.module,
      format: "es",
      sourcemap: true,
    },
  ],
  external: Object.keys(pkg.peerDependencies || {}),
  plugins: [
    resolve(),
    commonjs({
      exclude: [
        "src/**",
        "../../node_modules/lodash-es/**",
        "../../node_modules/symbol-observable/es/**",
      ],
      include: ["../../node_modules/**"],
      namedExports: {
        "../../node_modules/react-redux/node_modules/react-is/index.js": [
          "isValidElementType",
          "isContextConsumer",
        ],
        "../../node_modules/@material-ui/utils/node_modules/react-is/index.js": [
          "ForwardRef",
          "Memo",
        ],
        "../../node_modules/react-dom/index.js": [
          "findDOMNode",
          "unstable_batchedUpdates",
        ],
        "../../node_modules/react-is/index.js": [
          "isValidElementType",
          "isElement",
          "ForwardRef",
          "typeOf",
          "Memo",
          "isContextConsumer",
        ],
        "../../node_modules/prop-types/index.js": ["element", "elementType"],
      },
    }),
    babel({
      exclude: "../../node_modules/**",
    }),
    postcss(),
  ],
};
