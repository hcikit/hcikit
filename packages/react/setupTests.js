import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

let error = console.error;

console.error = function (message) {
  error.apply(console, arguments); // keep default behaviour
  throw message instanceof Error ? message : new Error(message);
};
