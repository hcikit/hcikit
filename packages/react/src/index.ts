import { PropsWithChildren, ReactElement, Component } from "react";
import { ControlFunctions } from "./core/Experiment.js";
import * as allTasks from "./tasks/index.js";

export * from "./tasks/index.js";
export * from "./components/index.js";

export { allTasks };

export { withGridItem } from "./GridLayout.js";

export { default as GridLayout } from "./GridLayout.js";
export { default } from "./core/Experiment.js";
export * from "./core/Experiment.js";
export {
  BasePersistence,
  StoragePersistence,
  FileSystemPersistence,
} from "./persistence/index.js";

type TaskProps = ControlFunctions;
type StaticProps = {
  layoutProps?: Record<string, any>;
};

export type Task<P = {}> = React.ComponentType<TaskProps & P> & StaticProps;

export interface FunctionTask<P = {}>
  extends React.FunctionComponent<P & ControlFunctions> {
  (
    props: PropsWithChildren<P & ControlFunctions>,
    context?: any
  ): ReactElement<any, any> | null;
}

export abstract class ClassTask<P = {}, S = {}, SS = any> extends Component<
  P & TaskProps,
  S,
  SS
> {}
