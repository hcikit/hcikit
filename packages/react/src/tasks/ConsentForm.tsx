import React from "react";
import Markdown from "markdown-to-jsx";

import {
  Button,
  Checkbox,
  FormControlLabel,
  FormControl,
  FormHelperText,
  FormGroup,
} from "@material-ui/core";

import { CenteredNicePaper } from "../components";
import { withGridItem } from "../GridLayout";
import { useState } from "react";
import { useExperiment } from "../core/Experiment";

export const ConsentForm: React.FunctionComponent<{
  content: string;
  questions: Array<{ label: string; required: boolean }>;
}> = ({ content, questions }) => {
  const [answers, setAnswers] = useState(
    questions.reduce<Record<string, boolean>>((prev, { label }) => {
      prev[label] = false;
      return prev;
    }, {})
  );

  const isFormComplete = () => {
    // Every value determines whether or not it is in the needed position. If it is optional, that is always true. Otherwise, it is the value of the current thing.
    const requiredNotFilled = questions.map((question) =>
      question.required ? answers[question.label] : true
    );

    // If any of these are false,
    return requiredNotFilled.every(Boolean);
  };

  const experiment = useExperiment();

  return (
    <CenteredNicePaper>
      <Markdown>{content}</Markdown>
      <FormControl required error={!isFormComplete()}>
        <FormGroup>
          {questions.map((question) => {
            return (
              <FormGroup key={question.label}>
                <FormControlLabel
                  key={question.label}
                  control={
                    <Checkbox
                      required={Boolean(question.required)}
                      onChange={(event) =>
                        setAnswers({
                          ...answers,
                          [question.label]: event.target.checked,
                        })
                      }
                      color="primary"
                    />
                  }
                  label={`${question.required ? "* " : ""}${question.label}`}
                />
              </FormGroup>
            );
          })}
          <FormHelperText>Required consent not given.</FormHelperText>

          <Button
            onClick={() => {
              if (isFormComplete()) {
                experiment.advance();
              }
            }}
            variant="contained"
            color="primary"
          >
            Submit
          </Button>
        </FormGroup>
      </FormControl>
    </CenteredNicePaper>
  );
};

export default withGridItem(ConsentForm, "task");
