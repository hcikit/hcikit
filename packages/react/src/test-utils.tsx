import { render, RenderResult } from "@testing-library/react";

import {
  ConfigContext,
  ConfigMutatorContext,
  ControlFunctions,
} from "./core/Experiment.js";
import { Configuration, ExperimentIndex } from "@hcikit/workflow";
import React from "react";
import { BasePersistence } from "./persistence/index.js";

class DoNothingPersistence extends BasePersistence {
  load(): Promise<Configuration | undefined> {
    return Promise.resolve(undefined);
  }
  save(state: Configuration): Promise<void> {
    return Promise.resolve();
  }
  clear(): Promise<void> {
    return Promise.resolve();
  }
  init(): Promise<void> {
    return Promise.resolve();
  }
}

const identities: ControlFunctions = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  advance: (_index?: ExperimentIndex) => {
    // do nothing.
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  log: (_log: unknown) => {
    // do nothing.
  },
  modify: (
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _modifiedConfig: Record<string, unknown>,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _index?: ExperimentIndex
  ) => {
    // do nothing.
  },
  persistence: new DoNothingPersistence(),
};

export const renderWithProvider = (
  Component: React.ReactElement,
  funcs: Partial<ControlFunctions> = identities,
  configuration: Configuration = {}
): RenderResult => {
  return render(
    <ConfigContext.Provider value={configuration}>
      <ConfigMutatorContext.Provider value={{ ...identities, ...funcs }}>
        {Component}
      </ConfigMutatorContext.Provider>
    </ConfigContext.Provider>
  );
};
