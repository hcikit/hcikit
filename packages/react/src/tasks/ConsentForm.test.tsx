import React from "react";
import ConsentForm from "./ConsentForm.js";
import { screen } from "@testing-library/react";
import { renderWithProvider } from "../test-utils.js";
import { jest } from "@jest/globals";
import userEvent from "@testing-library/user-event";

describe("ConsentForm", () => {
  it("renders without crashing", () => {
    renderWithProvider(
      <ConsentForm
        content="consent form"
        questions={[{ label: "I consent", required: true }]}
      />
    );
  });

  it("doesn't advance if consent not given", async () => {
    const advance = jest.fn();

    renderWithProvider(
      <ConsentForm
        content="consent form"
        questions={[{ label: "I consent", required: true }]}
      />,
      {
        advance,
      }
    );

    await userEvent.click(screen.getByText("Submit"));
    expect(advance).toBeCalledTimes(0);
    screen.getByText("Required consent not given.");
  });

  it("advances only when all consent is given", async () => {
    const advance = jest.fn();

    renderWithProvider(
      <ConsentForm
        content="consent form"
        questions={[
          { label: "I consent", required: true },
          { label: "I consent again", required: true },
        ]}
      />,
      {
        advance,
      }
    );

    await userEvent.click(screen.getByText("* I consent"));
    await userEvent.click(screen.getByText("Submit"));
    expect(advance).toBeCalledTimes(0);
    screen.getByText("Required consent not given.");

    await userEvent.click(screen.getByText("* I consent again"));
    await userEvent.click(screen.getByText("Submit"));
    expect(advance).toBeCalledTimes(1);
  });

  it("advances when optionals left blank", async () => {
    const advance = jest.fn();

    renderWithProvider(
      <ConsentForm
        content="consent form"
        questions={[
          { label: "I consent", required: true },
          { label: "I consent again", required: false },
        ]}
      />,
      {
        advance,
      }
    );

    await userEvent.click(screen.getByText("* I consent"));
    await userEvent.click(screen.getByText("Submit"));
    expect(advance).toBeCalledTimes(1);
  });

  it("advances when consent given", async () => {
    const advance = jest.fn();

    renderWithProvider(
      <ConsentForm
        content="consent form"
        questions={[{ label: "I consent", required: true }]}
      />,
      {
        advance,
      }
    );
    await userEvent.click(screen.getByText("* I consent"));
    await userEvent.click(screen.getByText("Submit"));
    expect(advance).toBeCalledTimes(1);
  });

  it("renders properly", () => {
    expect(
      renderWithProvider(
        <ConsentForm
          content={`
  # Title
  this is **bold**.
   - and
   - a
   - list
      `}
          questions={[
            { label: "I consent", required: true },
            { label: "I consent optional", required: false },
          ]}
        />
      ).asFragment()
    ).toMatchSnapshot();
  });
});
