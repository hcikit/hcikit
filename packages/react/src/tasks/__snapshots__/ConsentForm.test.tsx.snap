// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentForm renders properly 1`] = `
<DocumentFragment>
  <div
    style="grid-area: task;"
  >
    <div
      class="sc-gKPRtg cmBYij"
    >
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 sc-jSUZER goswre css-1ps6pg7-MuiPaper-root"
        >
          <h1>
            Title
          </h1>
          

          <p>
            this is 
            <strong>
              bold
            </strong>
            .
          </p>
          

          <ul>
            

            <li>
              and
            </li>
            

            <li>
              a
            </li>
            

            <li>
              list
            </li>
            

          </ul>
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <div
              class="MuiFormGroup-root Mui-error css-dmmspl-MuiFormGroup-root"
            >
              <div
                class="MuiFormGroup-root Mui-error css-dmmspl-MuiFormGroup-root"
              >
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd Mui-error css-j204z7-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                  >
                    <input
                      class="PrivateSwitchBase-input css-1m9pwf3"
                      data-indeterminate="false"
                      required=""
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                  >
                    * I consent
                  </span>
                </label>
              </div>
              <div
                class="MuiFormGroup-root Mui-error css-dmmspl-MuiFormGroup-root"
              >
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd Mui-error css-j204z7-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                  >
                    <input
                      class="PrivateSwitchBase-input css-1m9pwf3"
                      data-indeterminate="false"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                  >
                    I consent optional
                  </span>
                </label>
              </div>
              <p
                class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
              >
                Required consent not given.
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Submit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
